src/kernel/schedule.c

```diff
@@ -154,13 +145,93 @@ NOTRACE void __attribute__((noreturn)) runloop_internal()
+                    if (f == INVALID_ADDRESS) {
+                        sched_debug("migrating thread from idle CPU %d to self\n", cpu);
+                        f = dequeue(cpui->thread_queue);
+                        if (f != INVALID_ADDRESS)
+                            f[FRAME_QUEUE] = u64_from_pointer(ci->thread_queue);
+                    }
+                    if (!queue_empty(cpui->thread_queue))
+                        wakeup_cpu(cpu);
+                }
+            }
+
+            if (f == INVALID_ADDRESS) {
+                /* No threads found in idle CPUs: try to steal a thread from a
+                 * CPU that is currently running another thread. */
+                for (u64 cpu = ci->id + 1; ; cpu++) {
+                    if (cpu == total_processors)
+                        cpu = 0;
+                    if (cpu == ci->id)
+                        break;
+                    cpuinfo cpui = cpuinfo_from_id(cpu);
+                    if (cpui->state == cpu_user) {
+                        f = dequeue(cpui->thread_queue);
+                        if (f != INVALID_ADDRESS) {
+                            sched_debug("migrating thread from CPU %d to self\n", cpu);
+                            f[FRAME_QUEUE] = u64_from_pointer(ci->thread_queue);
+                            break;
+                        }
+                    }
+                }
+            }
+        } else {
+            for (u64 cpu = ci->id + 1; idle_cpu_mask; cpu++) {
+                /* Wake up idle CPUs that have a non-empty thread queue, and if
+                 * our thread queue is non-empty, migrate our threads to idle
+                 * CPUs. */
+                if (cpu == total_processors)
+                    cpu = 0;
+                if (cpu == ci->id)
+                    break;
+                if (!(idle_cpu_mask & (1 << cpu)))
+                    continue;
+                cpuinfo cpui = cpuinfo_from_id(cpu);
+                context fr;
+                if (!queue_empty(cpui->thread_queue)) {
+                    wakeup_cpu(cpu);
+                } else if ((fr = dequeue(ci->thread_queue)) != INVALID_ADDRESS) {
+                    sched_debug("migrating thread from self to idle CPU %d\n", cpu);
+                    fr[FRAME_QUEUE] = u64_from_pointer(cpui->thread_queue);
+                    enqueue(cpui->thread_queue, fr);
+                    wakeup_cpu(cpu);
+               }
+            }
+        }
+        if (f != INVALID_ADDRESS) {
+            if (!timer_updated) {
+                timestamp here = now(CLOCK_ID_MONOTONIC);
+                s64 timeout = ci->last_timer_update - here;
+                if ((timeout < 0) || (timeout > runloop_timer_max)) {
+                    sched_debug("setting CPU scheduler timer\n");
+                    runloop_timer(runloop_timer_max);
+                    ci->last_timer_update = here + runloop_timer_max;
+                }
+            }
+            t = (thunk)f[FRAME_RUN];
```

Comment:
Yes, for now the only reason I switched to enqueuing frames instead of thunks is to be able change the FRAME_QUEUE pointer when a thread is migrated; this way, we only change it when necessary, as opposed to setting it at every frame schedule in run_thread_frame(). And I thought this could be needed in the future to be able to access the relevant nanos_thread (via f[FRAME_THREAD]) to implement things such as CPU affinity.
But OK, we can stick to queuing thunks for now.